### DISCUSSION TITLE
Better, clearer maps

### DISCUSSION THREAD

**Discussion by JoeCohen** [MushroomObserver/mushroom-observer#2088] [Ideas]
What can do to generate comprehensible maps?
I'm attaching an email chain with power user Joanne Schwartz, who reports that her troubles with maps started recently.
She wanted to see a map of her Observations on Santa Cruz Island.
IMO the maps we generated are incomprehensible. 
- There's a mix of points and boxes 
- No explanation what the boxes mean
- Boxes covering an area different than what the user started with, and
- Points outside that area.

[Uploading Joanne maps question.pdf…]()

**Comment by nimmolo** [comment]
I feel like we discussed this and agreed on the following plan some time ago:

- Plot all Observation locations as points. 
- Most observations do not have point data (!!!), so use the center points of the Locations. 
- Use Google's [cluster renderer](https://developers.google.com/maps/documentation/javascript/marker-clustering) to aggregate coinciding points into multiples

The potential problem with this plan is that 
- The map depends on captions that we supply, to link directly to the actual observations
- Without some level of randomization, many center-points will plot as irreducible clusters of multiple observations. 
- I'm not at all sure that Google will let us give separate captions for coinciding locations. This is testable, though.

So, UI wise, it would be much easier to give the map points captions if we slightly randomized coinciding center-points (within a standard deviation equal to half the bounds of the box, say). But as @mo-nathan pointed out, this implies a level of "precision" that is fake and could be misleading. To remedy this, I proposed giving randomize-differentiated points a different color that signals that the location is approximate - I still kind of prefer this, because we will need that different color signal also for observations that _do_ have precise GPS but have "location hidden" by the user.

Mitigating consideration:
- You can always switch from the map view to a matrix box list. But it's not the same as zooming in to a map point and clicking the observation at that point.

**Reply by nimmolo** [reply]
Note my previous response edited on github - please refer to the full discussion on GitHub for the up-to-date post, rather than replying via email if possible.

**Comment by mo-nathan** [comment]
I like the idea of "box" locations having a different color.  It would also be reasonable in my view to have clicking on that location take the user to the set of observations in a matrix box mode if we can pull that off.  It might also be really helpful to give users a way to ignore "box" locations and only plot data that has real GPS values.

**Comment by JoeCohen** [comment]
Thanks Andrew. I couldn't find the previous discussion. That plan will go a long way to helping users. It's a difficult problem.

**Comment by JoeCohen** [comment]
It also seems like there are bugs now. For instance, Joanne
- Started with this Observation https://mushroomobserver.org/488916, of _Amanita muscaria subsp. flavivolvata,_ which is geolocated, and whose Location is [Santa Cruz Island, Santa Barbara Co., California, USA](https://mushroomobserver.org/locations/5059)
- clicked Occurence Map
- But the resulting map shows no occurences on Sata Cruz Island.

**Reply by nimmolo** [reply]
Hm. well, if we investigate the query for Names by location (in `Names::MapController#show`):
```ruby
      @query = create_query(:Observation, :all, names: @name.id)
      apply_content_filters(@query)
      @observations = @query.results(include: :location, limit: 10_000).
                      select { |o| o.lat || o.location }
```
What is it sorting by? 
```
   (278.1ms)  
      SELECT DISTINCT observations.id
      FROM `observations` 
      WHERE observations.name_id IN (5066)
  ORDER BY observations.when DESC, observations.id DESC
  LIMIT 10000
```
Seems to be `observations.when, observation.id DESC`... i would guess that would return obs 488916 in the first 10000, because there are not that many obs. And it does indeed:
```
Observation Load (14.3ms)  SELECT `observations`.* FROM `observations` WHERE `observations`.`id` IN (... 496336, 496335, 493965, 488916, 498083, 482169, ...)
```
So it should also be loading the location from that observation's location, whose ID is `5059`, and it's doing:
```
Location Load (4.3ms)  SELECT `locations`.* FROM `locations` WHERE `locations`.`id` IN (... 4059, 9126, 5059, 12880, 16810, ...)
```
So far, so good. To be continued, going to put some debuggers in `CollapsibleCollection`

**Reply by nimmolo** [reply]
This is as far as I've tracked it down: a raw "set" is created for Joanne's point location (GPS lat/lng, not a box). The `CollapsibleCollectionOfMappableObjects` class works by aggregating/collapsing each "set" with its neighbors into larger boxes, until a suitable level of precision is reached.

So her point is getting lost when aggregating the point into boxes, in `group_objects_into_sets`. It's probably something to do with the point not aggregating well with boxes. Continuing to investigate.

**Comment by nimmolo** [comment]
Welp, Jason figured this out pretty quickly after I spent a number of hours trying to debug it by tracing the handling of the location data through the iterations of boxes. Fresh eyes!

Here is what’s going on. In Joanne's case, the occurrence map of name 5066, "Amanita muscaria subsp. flavivolvata", there is one daggone obs (392296) with the vague location “USA”.

Because we do not order locations by area (and this is not easy to do…), during the process of collapsing the many obs of "Amanita muscaria subsp. flavivolvata" into a manageable number of boxes to display on the map, the mega-location “USA" has subsumed the next several specific locations within it, obscuring them. One of these is Joanne’s observation on “Santa Cruz Island”. If you click within the box for the whole USA, including Alaska and Hawaii, you’ll find the Santa Cruz Island observation.

This is no doubt happening on any map where there are vague observation locations, not only this map.

So I can think of two things we can do:

- In the mapsets that get sent to the maps, filter out not only “Earth”, but also all locations where the box size is greater than 1º. The argument being that these are useless anyway, for the purposes of mapping. (Although a 1º x 1º area is much bigger at the poles than the equator, Jason observed that this is offset by the fact that there are very few fungi living at the poles.)
- Add a validation for observations, that forces people to associate the obs with a more specific location if they pick one bigger than this size. This may be complicated to pull off, though, because "create obs" goes through “create location” and of course it’s possible to create general locations.

**Comment by pellaea** [comment]
Thought: instead of trying to prevent vague observations/locations, maybe
just a warning along the lines of: "This observation/location will not
appear on maps because it is too vague. If you would like it to be included
in maps, make this location smaller than 1 degree on a side or choose a
more precise location."

**Comment by nimmolo** [comment]
I like the idea of a warning. It could also come up during editing of an obs, i guess.

**Comment by pellaea** [comment]
Oh yeah, could happen client-side, too, at least in some cases.  If you
just select a location, it won't necessarily know if the location is vague
based on the autocomplete data (just name).

**Comment by nimmolo** [comment]
How about printing a warning (not a flash) in the show obs view at all times, if the location is nonspecific, that will explain the change in "policy"? We could vary the verbiage if it's the current user's own obs.

Public warning:
`This observation's location will not appear on maps because it is too vague.`
Addendum to obs owner:
`If you would like it to be included in maps, make this location smaller than 1 degree on a side, or choose a more precise location.`

I'm thinking it might spur people to clean up their own data, because it would appear even when reviewing old obs.

**Comment by JoeCohen** [comment]
IMO
1. It's okay to ask users to pick a different, more precise Location.
2. I'm concerned about the side-effects of modifying a Location. It may
- Put the lat/lon of other Observations outside the bounding box of the
modified Location.
- May otherwise surprise other Users who pick the modified Location for new
Observations.
- May induce users to narrow the bounding box so that it doesn't fit the
natural meaning of the Location.name.
3. The warning should include at least one example.

**Comment by pellaea** [comment]
Joe raises a good point about the verbage "make this location smaller".
People are already confused about how to edit an observation's location,
let's not further encourage people to modify locations!

But it seems like a win all around to gently encourage users to provide
useful location data.

**Comment by nimmolo** [comment]
Is everybody ok with the verbiage I put in the PR?

**Comment by pellaea** [comment]
Works for me.

**Comment by nimmolo** [comment]
### Proposal for maps of single observations: Start to favor GPS point data over Locations ###

PR #2092 makes a change in display for "Map of Observation" where the obs has "point" data: it assumes that the GPS info is more reliable than the associated Location record. If the point disagrees with the location, it displays that point, rather than defaulting to the Location bounding box, as currently. 

I'd like to hear how people feel about this change.

#### Background ####

Currently, when people create an observation:

- We ask people for a GPS point _and_ a Location (a database entry with a box)
- We autofill the last location used, although of course it may be wildly incorrect. If it disagrees with the GPS, we ask people to create a Location entry if there isn't one that fits, but they don't have to. (Check out observation [419427](https://mushroomobserver.org/419427), part of a foray in Idaho where the actual GPS saved on the obs is 50 miles from the associated Location.) Creating a Location is time consuming and we can safely assume many people skip this step, as this user obviously did. 
- We flash an alert that the Location is dubious if their GPS point lies outside a narrow delta from the Location box bounds. But if they go ahead and save the obs, we don't favor the GPS point when mapping that obs, we favor the Location.

Even though GPS is not 100% reliable either, it's the GPS point that should probably be considered definitive, not the Location the observation is associated with. I feel like the way we handle all this currently adds unnecessary confusion and effort for users.

Also, the tight check for `dubious?` may get in the way of simplifying the process of creating an observation. In remote or rural places, and especially on public land, the boundary of the common-sense "postal location" is often some distance away from the coordinates of the obs. But together with the GPS, it's likely "good enough" to identify the observation location. 

So this PR also loosens the requirement for `lat_long_close?`. I'm thinking particularly if we let Google look up postal addresses from GPS coordinates in the future, we'll probably want to let the returned address be "the location", with no further ado - and possibly even "auto-create" an MO location from that public map data. When I spot-checked some rural coordinates with GMaps reverse-lookup, many of the locations returned were "dubious" by our current standards — the point lay well outside our "delta" from the location boundaries.

It's GPS data we want to encourage people to provide, not "better" entries in the Location table, in my opinion.

**Reply by nimmolo** [reply]
#### What we show when GPS and Location diverge: ####

Observation 419427 is part of a group of observations by a single user that were all associated with one Location (15788, Sage Hen Mountain, Idaho), possibly because it kept being autofilled as their "last location used" in the Create Obs form.  This observation also has GPS point data, and according to that, it seems to have been found ~50mi from the location assigned to it.

Currently on main, it maps the location here:
 
![Screen Shot 2024-04-11 at 10 30 37 PM](https://github.com/MushroomObserver/mushroom-observer/assets/1948095/730282ad-a00f-452e-99a6-23ed104b75dd)

The PR branch shows the GPS location:

![Screen Shot 2024-04-11 at 10 28 57 PM](https://github.com/MushroomObserver/mushroom-observer/assets/1948095/c52182c7-ee22-41fa-bf4e-198f950f2a90)

Distance between these places is pretty great:

![Screen Shot 2024-04-11 at 10 32 41 PM](https://github.com/MushroomObserver/mushroom-observer/assets/1948095/da95ac24-8312-46d9-af11-86fea3cbdfff)

**Reply by nimmolo** [reply]
(Discussion has been [edited for clarity](https://github.com/MushroomObserver/mushroom-observer/discussions/2088 ))

**Comment by nimmolo** [comment]
As Terri points out, occurrence maps are currently box locations only, to avoid the mixed “pins and boxes” display people were complaining about.

> I created this new obs with GPS coordinates but it still has a message that the location is too vague:  [Mushroom Observer: Observation 548898: Melanoleuca Pat.](https://mushroomobserver.org/548898)  I would like all of my vouchered sequenced obs to show on MO occurrence maps if possible but there are over 1000 and I don't have the time to edit each one if it requires lots of steps.  I would have thought adding GPS coordinates would do it.  Any ideas?

These are our choices for occurrence maps: show boxes and only boxes, or show pins and only pins, with Google’s clustering. I believe people are in favor of the switch to pins, but please do speak up if you foresee problems with this approach.

- Currently, all users must give the obs an associated Location if they want it to appear on occurrence maps. Like Terri says, it’s time consuming. But it makes the maps clearer, assuming they have boxes.

- We have to write all the code to switch occurrence maps to pins. Box locations would instead show a pin at the center point. The viability of this depends on whether or not google will give separate captions for co-incident points, yet to be investigated.

**Comment by mo-nathan** [comment]
Generally I like the direction of making GPS coordinates take priority and doing something reasonable for observations that we only have location information for.  However, note that hidden observations need to be handled correctly.  For the new "hidden locations" I essentially rounded the GPS coordinates for the location to the nearest 0.1 degrees (round 6 miles in most reasonable places in the world).  There's an interesting case of what to do if you have a hidden observation with GPS in a hidden location that the user does not have permission to see.  I think the right thing to do in this case is to use the hidden location.  I think it would be reasonable to not show the precise location even if the person doing has permission to see the precise location.

**Reply by nimmolo** [reply]
That sounds about right. But I want to be sure everyone realizes this is an _either-or choice_, and is on board with it, before i put a lot of work into it. 

## As far as I know, we cannot have a mix of boxes and points on a map with Google point clustering. ##

This means NO MORE BOXES on occurrence maps. If we switch to automatic point clustering, for observations without GPS, or observations with hidden GPS, or hidden Locations, or just Locations, we have to make the location boil down to a point. We can give it a different color if it's a Location center point, and a third color if it's hidden. But it's got to be represented by a point, and that may confuse some people.

**Comment by JoeCohen** [comment]
Thanks for the reminder of where we are, where we're going, and prohibition
on mixing boxes and points.
I'm good with no boxes.
I'm good with different color pins for (1) obs with public lat/lon, (2) all
other obs. IMO it gets too confusing to have to more colors, but I'm also
okay with that if you to go that way.
Even better (but maybe something for another PR) would be different
style pins, such as (1) regular pin for obs with public gps, (2) circle for
obs with obscured gps or location, (3) quadrilaterial pin for obs with only
public location.
See
https://cloudfresh.com/en/blog/reimagine-your-mapping-experience-with-google-maps-advanced-markers/
,
https://developers.google.com/maps/documentation/javascript/reference/advanced-markers
-Joe

**Comment by mo-nathan** [comment]
I'm perfectly happy getting rid of boxes.  They were a short cut I took a long time ago to avoid doing real polygons and I doubt anyone is really attached to them.  We may get some grumbles from people who don't want their cheese moved, but I would be interested in hearing from them what they actually like about it.

Little MO mushroom pins would be cool.  But i wouldn't be in favor of the generic Mario-esque unicode mushroom (🍄).  However, I'd also be perfectly happy with colored pins or essentially boxes and circles as Joe suggests.  INat's square heat map approach is also nice, but probably a lot more work.

**Comment by pellaea** [comment]
FWIW, I'm also fine with removing boxes altogether.  Centerpoint of most
box locations is just fine.  Locations too vague to include on the map are
omitted (as they already are, I think).

**Comment by JoeCohen** [comment]
Re pin shape: I like the narrowest possible (vs little MO mushroom pins) because they obscure less of the map (and other pins).

**Reply by nimmolo** [reply]
We may get a choice on _pins_, but i'm not sure we get a choice on _cluster markers_.

**Comment by pellaea** [comment]
Here are some "pins" (just dots really) that I created for my own private
herbarium database system.  A bit boring, but they may meet Joe's
preferences?
